modules/imgcodecs/src/loadsave.cpp

```diff
@@ -1032,6 +923,265 @@ bool haveImageWriter( const String& filename )
+
+void ImageCollection::Impl::reinit() {
+    m_current = 0;
+#ifdef HAVE_GDAL
+    if (m_flags != IMREAD_UNCHANGED && (m_flags & IMREAD_LOAD_GDAL) == IMREAD_LOAD_GDAL) {
+        m_decoder = GdalDecoder().newDecoder();
+    }
+    else {
+#endif
+    m_decoder = findDecoder(m_filename);
+#ifdef HAVE_GDAL
+    }
+#endif
+
+    m_decoder->setSource(m_filename);
+    m_decoder->readHeader();
+}
+
+Mat ImageCollection::Impl::at(int index) {
+    if(index < 0 || (size_t)index >= m_size)
+        throw cv::Exception(1, "Index out of bounds",
+                            String("ImageCollection::at"),
+                            __FILE__,
+                            __LINE__);
+    return operator[](index);
+}
+
+Mat ImageCollection::Impl::operator[](int index) {
+    if(m_pages[index].empty()) {
+        // We can't go backward in multi images. If the page is not in vector yet,
+        // go back to first page and advance until the desired page and read it into memory
+        if(m_current != index) {
+            reinit();
+            for(int i = 0; i != index && advance(); ++i) {}
+        }
+        m_pages[index] = read();
+    }
+    return m_pages[index];
+}
+
+void ImageCollection::Impl::release(int index) {
+    if(index < 0 || (size_t)index >= m_size)
+        throw cv::Exception(1, "Index out of bounds",
+                            String("ImageCollection::at"),
+                            __FILE__,
+                            __LINE__);
+    m_pages[index].release();
+}
+
+/* ImageCollection API*/
+
+ImageCollection::ImageCollection() : pImpl(new Impl()) {}
+
+ImageCollection::ImageCollection(const std::string& filename, int flags) : pImpl(new Impl(filename, flags)) {}
+
+void ImageCollection::init(const String& img, int flags) { pImpl->init(img, flags); }
+
+size_t ImageCollection::size() const { return pImpl->size(); }
+
+Mat ImageCollection::at(int index) { return pImpl->at(index); }
+
+Mat ImageCollection::operator[](int index) { return pImpl->operator[](index); }
+
+void ImageCollection::release(int index) { pImpl->release(index); }
+
+Ptr<ImageCollection::Impl> ImageCollection::getImpl() { return pImpl; }
+
+/* Iterator API */
+
+ImageCollection::iterator ImageCollection::begin() { return pImpl->begin(this); }
+
+ImageCollection::iterator ImageCollection::end() { return pImpl->end(this); }
+
+ImageCollection::iterator::iterator(ImageCollection* col) : m_pCollection(col), m_curr(0) {}
+
+ImageCollection::iterator::iterator(ImageCollection* col, int end) : m_pCollection(col), m_curr(end) {}
+
+Mat ImageCollection::iterator::operator*() {
+    if(m_pCollection == nullptr)
+    {
+        throw std::runtime_error("Invalid image collection");
+    }
+    return m_pCollection->getImpl()->operator[](m_curr);
+}
+
+ImageCollection::iterator& ImageCollection::iterator::operator++() {
+    if(m_pCollection->pImpl->currentIndex() == m_curr) {
+        m_pCollection->pImpl->advance();
```

Comment:
do we need to call advance() here? Can operator [] take care of advancing internal pointer?

Reply:
@ocpalo The question is still open. I suspect, that it's a bug.

Reply:
Vadim also suspected that this is a bug. But I do not think so. He requested a test to break this and I implemented it and it seems it works fine. I did not included in the comment because it was not well-written test. If you think a specific scenario that can break this, please tell me. So I can included in the unit tests and see if its broken or not.

The idea is, only iterators can advance the to next page that satisfy the requirement which is iterator internal counter is equal to current page index. Since iterators should not pass `end()`, So if this requirement is not satisfied, it does not call `nextPage()`. So file status is protected. Iterator count is incremented in any case, which is required.

I will write a detailed description of how that incrementing index works tonight.